# Python Part II: Intro to Pandas

In Python Part I, we flew through introductory Python concepts, everything from the `print()` statement to For Loops, and used that information to write cool little applications like simulating a coin flip. At the end, we covered packages which are importable Python obejcts with defined methods and attributes. We looked at the `math` package, we includes relevant mathmatical functions like `log10()` and `gcd()`. We also looked at the `random` package, with allowed us to create random variables. 

We will now introduce the Pandas package, which is a Python package that allows users easily to manipulate datasets in Python.

*Note: In the walkthrough below, the first instance of each new Pandas method or attribute will be hyperlinked directly to the Pandas documentation for that method/attribute. Feel free to reference that documentation for additional information.*

# Part 1: The DataFrame

First, to start using Pandas, we have to import it. Typically, Pandas is imported as follows

In [70]:
import pandas as pd

The above code imports Pandas and then assigns the package a nickname, in this case `pd`. Therefore, whenever we refer to Pandas we can just type `pd`. 

In the below code, we use Pandas to read in our first dataset, specifically the `sales.csv` file stored in the "Intro to Python - II" repository. 

In [71]:
sales_csv = r"https://raw.githubusercontent.com/cra-international/Intro-to-Python/master/Intro%20to%20Python%20-%20II/cookie_sales.csv"
sales_df = pd.read_csv(sales_csv)

In the above code, we reference the Pandas package using the `pd` notation discussed above. We then call the [`read_csv()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) function, which has one *required* argument, in this case the path to the file. As the documentation notes, "Any valid string path is acceptable. The string could be a URL." So in this case, we pass a URL to the function which itself points to a CSV file.

Great! So, what was returned by this `read_csv()` function? Let's find out! We can use Python's built-in `type()` function to find out.

In [72]:
print(type(sales_df))

<class 'pandas.core.frame.DataFrame'>


In this case, what was returned was a *DataFrames*, which is a data type we haven't seen before. That's because DataFrames are defined by the Pandas package. The majority of this demo will focus on learning about DataFrames, as this is where all the magic happens. DataFrames not only store data, but they also have *methods* that allow us to manipulate the data easily. We'll cover the methods in more detail, and exactly how DataFrames store data, below. 

### Now you try! 

Create a new dataframe, called `price_df`, that takes the value of the data stored in the `cookie_prices.csv` file. The URL to that file has already been stored in the `price_url` variable below. 

In [73]:
price_url = r"https://raw.githubusercontent.com/cra-international/Intro-to-Python/master/Intro%20to%20Python%20-%20II/cookie_prices.csv"
price_df = pd.read_csv(price_url)

## Part 2: Exploring Data

### <ins>The `.shape` Attribute</ins>

Now that we can read in the data, the next step is to explore the data and see what it looks like. So let's learn our first Pandas function. We can get a quick glimpse of the data by using the [`.head()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html) method, which by default prints the first five observations of the dataset.

In [74]:
sales_df.head()

Troop Number                    Box Sold Notes
0           179             Thin Mints   43   NaN
1           179        Caramel deLites   16   NaN
2           179  Peanut Butter Patties   22   NaN
3           179     Girl Scout S'mores   11   NaN
4           179              Lemonades   34   NaN

Here we see that there are 4 columns - Troop Number, Cookie, state, Boxes Sold, and Notes. Based on the looks of it, this looks like Girl Scout Cookie Sales data by troop and cookie type.

We can also look at the shape of the data using the [`.shape`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html) attribute. We have not covered attributes up to this point, but an *attribute* is a static value associated with a Python object. Like methods, attributes are accessed using the dot operator, however they are not followed by any parentheses. In this case, the `.shape`, attribute of a Pandas DataFrame returns a *tuple* of the number of observations and number of columns in the DataFrame, in that order. We will talk more about tuples below.

In [75]:
print(sales_df.shape)

(51, 4)


Based on this, we can see that the first value of the tuple is `40` and the second is `4`, indicating that there are 40 observations and 4 columns in `sales_df`. Tuples are very simlar to lists as they are ordered and therefore can be accessed using positional indexing. For example, we can get just the number of observations of `sales_df` by doing the following

In [76]:
print("sales_df has", sales_df.shape[0] , "observations.")

sales_df has 51 observations.


The key difference between tuples and lists, however, is that tuples are *not* mutable, meaning you can not change the values of a tuple using positional indexing.

### Now you try!

In the first code cell below, use the `.head()` method to look at the `price_df` variable you created above. What does the `price_df` dataset look like? What kind of data is it? Then, in the second code cell, print the number of columns of the data to the console by using positional indexing to reference the second value of the tuple returned by `.shape`.

In [77]:
price_df.head()

Cookie Dollars
0             Thin Mints       4
1        Caramel deLites       4
2  Peanut Butter Patties       4
3     Girl Scout S'mores       4
4              Lemonades       4

In [78]:
print("price_df has", price_df.shape[1], "columns.")

price_df has 2 columns.


### <ins>Pandas Data Types</ins>

DataFrames, like other 2-dimensional data storage types such as SQL tables or Stata datasets, defines data types as the *column* level. Therefore, like regular Python, Dataframes also have basic data types we need to learn about. These are 
* int64
* float64
* object
* bool
* datetime64

You should notice some similarities between these data types and the basic Python datatypes. `int64` is Pandas' integer data type, where the 64 denotes how many bits each value is comprised of. Since there are 8 bits in a byte, that means the largest value an int64 can be is $2^{64}$, which is typically plenty large enough. `float64` is simlar, but for floats and `bool` is a column of boolean values. 

How are strings stored? Strings are stored using the `object` data type. It's important to note that while all string columns are stored as `object`s, not all `object` columns are strings. For example, we may have a column of *mixed* values, say strings and numbers. That type of column will be cast as an `object` type, essentially converting the numbers in column to text.

Finally, there is one type of datatype we haven't encountered before, and that's the `datetime64` type. `datetime` objects are built-in Python storage objects that allow for easy creation, manipulation, and storage of dates. Pandas' `datetime64` data type is a vectorized form of of the `datetime` object type which allocates 8 bytes to each `datetime` value in the column. 

So, how do we see what data types specific columns are in our data? We can do that by referencing the [`.dtypes`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dtypes.html) attribute of a DataFrame.

In [79]:
sales_df.dtypes

Troop Number     int64
Box             object
Sold            object
Notes           object
dtype: object

This DataFrame attribute simply returns as text the data type of each column in the associated DataFrame.

Based on the first 5 observations of the data we saw from calling the `.head()` method, it's a bit odd that "Boxes Sold", which looks numeric, is of the `object` type. What's going on? Let's print the whole dataset to see. To do that in a Jupyter Notebook, we simply run a block of code with just the variable name of the DataFrame we want to see, as below

In [80]:
sales_df

Troop Number                     Box      Sold  \
0            179              Thin Mints        43   
1            179         Caramel deLites        16   
2            179   Peanut Butter Patties        22   
3            179      Girl Scout S'mores        11   
4            179               Lemonades        34   
5            179  Peanut Butter Sandwich         8   
6            179              Shortbread        38   
7            179            Thanks-A-Lot        26   
8            179  Caramel Chocolate Chip        17   
9            179           Toffee-tastic        29   
10           177              Thin Mints       117   
11           177         Caramel deLites        84   
12           177   Peanut Butter Patties        39   
13           177      Girl Scout S'mores        33   
14           177               Lemonades        56   
15           177  Peanut Butter Sandwich        12   
16           177              Shortbread        86   
17           177            Thanks-A-Lot        24   
18           177  Caramel Chocolate Chip        45   
19           177           Toffee-tastic        77   
20           177               Donations        20   
21           143              Thin Mints        94   
22           143         Caramel deLites        99   
23           143   Peanut Butter Patties  No Sales   
24           143      Girl Scout S'mores        76   
25           143               Lemonades       125   
26           143  Peanut Butter Sandwich  No Sales   
27           143              Shortbread        36   
28           143            Thanks-A-Lot        15   
29           143  Caramel Chocolate Chip        69   
30           143           Toffee-tastic       103   
31           212              Thin Mints        21   
32           212         Caramel deLites         7   
33           212   Peanut Butter Patties         3   
34           212      Girl Scout S'mores         5   
35           212               Lemonades        15   
36           212  Peanut Butter Sandwich         2   
37           212              Shortbread        17   
38           212            Thanks-A-Lot         9   
39           212  Caramel Chocolate Chip        12   
40           212           Toffee-tastic        13   
41           179              Thin Mints        43   
42           179         Caramel deLites        16   
43           179   Peanut Butter Patties        22   
44           179      Girl Scout S'mores        11   
45           179               Lemonades        34   
46           179  Peanut Butter Sandwich         8   
47           179              Shortbread        38   
48           179            Thanks-A-Lot        26   
49           179  Caramel Chocolate Chip        17   
50           179           Toffee-tastic        29   

                                    Notes  
0                                     NaN  
1                                     NaN  
2                                     NaN  
3                                     NaN  
4                                     NaN  
5                                     NaN  
6                                     NaN  
7                                     NaN  
8                                     NaN  
9                                     NaN  
10                                    NaN  
11                                    NaN  
12                                    NaN  
13                                    NaN  
14                                    NaN  
15                                    NaN  
16                                    NaN  
17                                    NaN  
18                                    NaN  
19                                    NaN  
20    *A kind soul just dropped off a $20  
21                                    NaN  
22                                    NaN  
23  *Severe Peanut Allergy Impacted Sales  
24                                    NaN  
25                                    NaN  
26  *Se

As we can see, while "Boxes Sold" is almost always numeric, there are two instances where it takes the value "No Sales". (That seems to be because, based on the "Notes" column, a "Severe Peanut Allergy Impacted Sales".) Those two string values have caused the whole column to be cast as an `object` type. 

There also seem to be some other problems with the data. For example, it looks like there are some duplicates of Troop 179's data at the bottom of the dataset. We will address how to resolve this and the data type problem later on in the walkthrough.


### Now You Try!

Print the data types of the `price_df` dataset below and the print the entire dataset to the console.

In [81]:
print(price_df.dtypes)
price_df

Cookie     object
Dollars    object
dtype: object


Cookie Dollars
0               Thin Mints       4
1          Caramel deLites       4
2    Peanut Butter Patties       4
3       Girl Scout S'mores       4
4                Lemonades       4
5    Peanut Butter Sanwich       4
6               Shortbread       4
7             Thanks-A-Lot       4
8   Caramel Chocolate Chip  $5.00 
9                Lemonades       4
10           Toffee-tastic       4

## Part 3: Renaming Columns


The `sales_df` DataFrame has four columns - "Troop Number", "Box", "Sold", and "Notes". Some of these column names, however, are not that informative. For example, we may want to change "Box" to "Cookie Type" and "Sold" to "Boxes Sold". Let's do that by using the [`.rename()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html) method, as below and then print the first five observations to see the results

In [82]:
sales_df = sales_df.rename(columns={"Box" : "Cookie Type",
                                    "Sold" : "Boxes Sold"})
sales_df.head()

Troop Number            Cookie Type Boxes Sold Notes
0           179             Thin Mints         43   NaN
1           179        Caramel deLites         16   NaN
2           179  Peanut Butter Patties         22   NaN
3           179     Girl Scout S'mores         11   NaN
4           179              Lemonades         34   NaN

We can now see that the column "Box" was changed to "Cookie Type" and "Sold" was changed to "Boxes Sold". 

The `.rename()` method can take multiple arguments, but the one most frequently used is the `columns` argument. Note that we haven't seen this type of argument notation before, but specific arguments can be specified using the following syntax `<argument name>=<value>`. Arguments specified in this way have a specific name, *keyword arguments*. All of Pandas' more complex methods rely on keyword arguments, so we will see this notation pop up more as we go along.

In this case, we specify the keyword argument `columns` within the call to the `.rename()` method and pass a *dictionary* as the value of the argument. The dictionary is structured so that the *key* is the existing column name and the *value* is the new column name. This dictionary contains multiple key-value pairs corresponding to two column name changes we intend to make. The result of the `.rename()` method is a new DataFrame with the updated column names. In this case, we update the `sales_df` variable by assigning that new DataFrame to the existing variable `sales_df`. 

### Now You Try!

The price variable in the `price_df` data is currently just called "Dollars". Rename that variable to "Price (\\$)" to make it more specific/informative. 

*Note: Pandas allows the use of special characters in column names!*

In [83]:
price_df = price_df.rename(columns={"Dollars" : "Price ($)"})
price_df.head()

Cookie Price ($)
0             Thin Mints         4
1        Caramel deLites         4
2  Peanut Butter Patties         4
3     Girl Scout S'mores         4
4              Lemonades         4

## Part 4: Indexing Data

How do you select just the data you want to work with? That can be done in Pandas using *indexing*, similar to how we used indexing to select specific values from lists in Intro to Python Part I. We will talk through all of the basics of indexing below as this is a crucial aspect of working with data in Pandas.

## <ins>The Index</ins>

In each print out of the datasets above, there is an additional "column" on the left-hand side of the data that is yet to be explained. This is called the DataFrame *index* and is a unique ID associated with each observation in the dataset. The DataFrame is very similar to a list in the sense that its (default) index is zero-indexed and can be used to access specific data within the Python object. 

Unlike lists, however, observations are accessed with a specific DataFrame attribute called [`iloc`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iloc.html), which stands for "index locate". Let's see how `.iloc` works below by printing `sales_df.iloc`. 

In [84]:
print(sales_df.iloc)

We see abvove that `sales_df.iloc` is an "iLocIndexer object", an object defined by the Pandas package. (That gobbledy-gook on the far right is the location of the object in memory in hexadecimal). You don't really need to worry much about this, but the most import part is that the "iLocIndexer object" can be indexed in the exact same way as a list. For example, we can access the first 5 observations of a DataFrame as follows

In [85]:
sales_df.iloc[0:5]

Troop Number            Cookie Type Boxes Sold Notes
0           179             Thin Mints         43   NaN
1           179        Caramel deLites         16   NaN
2           179  Peanut Butter Patties         22   NaN
3           179     Girl Scout S'mores         11   NaN
4           179              Lemonades         34   NaN

Perfect! This gives us information identical to `sales_df.head()`. You can also get multiple specific non-sequential rows by passing a *list* of the desired rows within the square brackets.

In [86]:
sales_df.iloc[[1, 3, 5]]

Troop Number             Cookie Type Boxes Sold Notes
1           179         Caramel deLites         16   NaN
3           179      Girl Scout S'mores         11   NaN
5           179  Peanut Butter Sandwich          8   NaN

What happens if you access just one row? Let's see!

In [87]:
print(sales_df.iloc[0])
print()
print(type(sales_df.iloc[0]))

Troop Number           179
Cookie Type     Thin Mints
Boxes Sold              43
Notes                  NaN
Name: 0, dtype: object

<class 'pandas.core.series.Series'>


We see that if we print just one observation, we get a print out of the individual values of the row as well as the "Name" or index of the row. If we check the type of the row using the `type()` function, we see that we no longer have a DataFrame but now have a Pandas [`Series`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html). 

Pandas stores individual rows or individual columns as Series, which you can think of as complex lists that have very helpful mathematical properties and defined methods. For example, the sum of two Series is not one longer Series, as it would be for lists, but rather the sum of two vectors as defined by vector algebra. 

*Note: You can print an empty line in Python by just calling* `print()`

### Now You Try!

Combine what you just learned with your knowledge of list-based indexing to print the last 20 observations of the `sales_df` dataset.

In [88]:
sales_df[-20:]

Troop Number             Cookie Type Boxes Sold Notes
31           212              Thin Mints         21   NaN
32           212         Caramel deLites          7   NaN
33           212   Peanut Butter Patties          3   NaN
34           212      Girl Scout S'mores          5   NaN
35           212               Lemonades         15   NaN
36           212  Peanut Butter Sandwich          2   NaN
37           212              Shortbread         17   NaN
38           212            Thanks-A-Lot          9   NaN
39           212  Caramel Chocolate Chip         12   NaN
40           212           Toffee-tastic         13   NaN
41           179              Thin Mints         43   NaN
42           179         Caramel deLites         16   NaN
43           179   Peanut Butter Patties         22   NaN
44           179      Girl Scout S'mores         11   NaN
45           179               Lemonades         34   NaN
46           179  Peanut Butter Sandwich          8   NaN
47           179              Shortbread         38   NaN
48           179            Thanks-A-Lot         26   NaN
49           179  Caramel Chocolate Chip         17   NaN
50           179           Toffee-tastic         29   NaN

### <ins>Column and Combined Indexing</ins>

It's important to note that in a DataFrame, not just observations have indices. Columns have indices as well, it just happens that when we read in our data, Pandas didn't use its default indices (0, 1, 2, 3, etc.), but gaved them *named indices*, in this case the column names stored in the first row of our csv data. Therefore, we can use the columns names to directly access specific columns by name using dictionary-like key indexing, as below

In [89]:
print(sales_df["Boxes Sold"])
print()
print(type(sales_df["Boxes Sold"]))

0           43
1           16
2           22
3           11
4           34
5            8
6           38
7           26
8           17
9           29
10         117
11          84
12          39
13          33
14          56
15          12
16          86
17          24
18          45
19          77
20          20
21          94
22          99
23    No Sales
24          76
25         125
26    No Sales
27          36
28          15
29          69
30         103
31          21
32           7
33           3
34           5
35          15
36           2
37          17
38           9
39          12
40          13
41          43
42          16
43          22
44          11
45          34
46           8
47          38
48          26
49          17
50          29
Name: Boxes Sold, dtype: object

<class 'pandas.core.series.Series'>


Like with the individual row we accessed using the `iloc` object, here accessing just one column returns a Series object. Also, like with `iloc` indexing, we can access multiple columns by passing a list of column names as below

In [90]:
sales_df[["Cookie Type", "Boxes Sold"]]

Cookie Type Boxes Sold
0               Thin Mints         43
1          Caramel deLites         16
2    Peanut Butter Patties         22
3       Girl Scout S'mores         11
4                Lemonades         34
5   Peanut Butter Sandwich          8
6               Shortbread         38
7             Thanks-A-Lot         26
8   Caramel Chocolate Chip         17
9            Toffee-tastic         29
10              Thin Mints        117
11         Caramel deLites         84
12   Peanut Butter Patties         39
13      Girl Scout S'mores         33
14               Lemonades         56
15  Peanut Butter Sandwich         12
16              Shortbread         86
17            Thanks-A-Lot         24
18  Caramel Chocolate Chip         45
19           Toffee-tastic         77
20               Donations         20
21              Thin Mints         94
22         Caramel deLites         99
23   Peanut Butter Patties   No Sales
24      Girl Scout S'mores         76
25               Lemonades        125
26  Peanut Butter Sandwich   No Sales
27              Shortbread         36
28            Thanks-A-Lot         15
29  Caramel Chocolate Chip         69
30           Toffee-tastic        103
31              Thin Mints         21
32         Caramel deLites          7
33   Peanut Butter Patties          3
34      Girl Scout S'mores          5
35               Lemonades         15
36  Peanut Butter Sandwich          2
37              Shortbread         17
38            Thanks-A-Lot          9
39  Caramel Chocolate Chip         12
40           Toffee-tastic         13
41              Thin Mints         43
42         Caramel deLites         16
43   Peanut Butter Patties         22
44      Girl Scout S'mores         11
45               Lemonades         34
46  Peanut Butter Sandwich          8
47              Shortbread         38
48            Thanks-A-Lot         26
49  Caramel Chocolate Chip         17
50           Toffee-tastic         29

What if you want to index a dataset based of of both columns and observations? Say we want to see only the first 10 observations of the "Cookie Type" and "Boxes Sold" columns? In that case, we'd use the [`.loc`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html) indexer, which indexes a DataFrame based on the name of its indices.

In [91]:
sales_df.loc[:10, ["Cookie Type", "Boxes Sold"]]

Cookie Type Boxes Sold
0               Thin Mints         43
1          Caramel deLites         16
2    Peanut Butter Patties         22
3       Girl Scout S'mores         11
4                Lemonades         34
5   Peanut Butter Sandwich          8
6               Shortbread         38
7             Thanks-A-Lot         26
8   Caramel Chocolate Chip         17
9            Toffee-tastic         29
10              Thin Mints        117

In the above case, we take advantage of the fact that the *name* of our row/observation index is the same as its position, which is the case by default. Therefore, we specify our doubly-indexed DataFrame by specifying the observations we want first (in this case by specifying a *slice* or range of observations), followed by a comma, followed by a *list* of our desired columns. 

### Now You Try!

Use the `.loc` indexer to access the 1st, and 21st, and 27th observations and just the "Troop Number" and "Notes" columns.

In [92]:
sales_df.loc[[0, 20, 26], ["Troop Number", "Notes"]]

Troop Number                                  Notes
0            179                                    NaN
20           177    *A kind soul just dropped off a $20
26           143  *Severe Peanut Allergy Impacted Sales

## <ins>Boolean Indexing</ins>

While good to know, it's sometimes not useful/pragmatic to access indiviual rows of a DataFrame using positional indexing. More often than not, it's more useful to be able to access rows based on a characteristic they share, such a specific value or condition. Indexing a DataFrame in that way is possible using boolean indexing.

For example, you might want to look at all data for Troop 177 Only. Rather identify the position of each observation assocaite with Troop 177 and then selecting observations using positional indexing, we can directly index by doing the following

In [93]:
sales_df[sales_df["Troop Number"] == 177]

Troop Number             Cookie Type Boxes Sold  \
10           177              Thin Mints        117   
11           177         Caramel deLites         84   
12           177   Peanut Butter Patties         39   
13           177      Girl Scout S'mores         33   
14           177               Lemonades         56   
15           177  Peanut Butter Sandwich         12   
16           177              Shortbread         86   
17           177            Thanks-A-Lot         24   
18           177  Caramel Chocolate Chip         45   
19           177           Toffee-tastic         77   
20           177               Donations         20   

                                  Notes  
10                                  NaN  
11                                  NaN  
12                                  NaN  
13                                  NaN  
14                                  NaN  
15                                  NaN  
16                                  NaN  
17                                  NaN  
18                                  NaN  
19                                  NaN  
20  *A kind soul just dropped off a $20

In the code above, rather than specify specific indices associated with the observations we want to isolate, we instead pass a comparison, in this case a comparison of a specific column against a specific value. In Python Part I, we saw that the result of a comparison like `5 < 6` is a boolean value. What does this comparison of a Pandas Series against a value return? 

In [94]:
sales_df["Troop Number"] == 177

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10     True
11     True
12     True
13     True
14     True
15     True
16     True
17     True
18     True
19     True
20     True
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
32    False
33    False
34    False
35    False
36    False
37    False
38    False
39    False
40    False
41    False
42    False
43    False
44    False
45    False
46    False
47    False
48    False
49    False
50    False
Name: Troop Number, dtype: bool

A comparison involving an individual column returns another Series, in this case a Series of type `bool`. Specifically, this series takes the value `True` whenever the compairson is True and `False` otherwise. In Pandas, this kind of Series that references specific observations using boolean values is called a *boolean mask*. A boolean mask, then, can be used to *boolean index* the dataset as a boolean value of `True` indiciates that we'd like to select the observation. 

### Now You Try!

Use boolean indexing to select just the observations from `sales_df` associated with "Toffee-tastic" cookies.

In [95]:
sales_df.loc[sales_df["Cookie Type"] == "Toffee-tastic"]

Troop Number    Cookie Type Boxes Sold Notes
9            179  Toffee-tastic         29   NaN
19           177  Toffee-tastic         77   NaN
30           143  Toffee-tastic        103   NaN
40           212  Toffee-tastic         13   NaN
50           179  Toffee-tastic         29   NaN

## Part 5: Cleaning Data

Up to this point, we have read in data, explored it, renamed columns (or "edited metadata" if you want to be fancy), and learned how to select just the data we want to work with. But now we will cover how you clean data so that we can finally analyze it.

Cleaning data is a broad umbrella, but can include changing variable types, dropping superfluous columns and unneeded observations, removing duplicate data, and much more. We will look into how to do many of these tasks below.

### <ins>Replace Observations</ins>

When we looked at the sales data earlier, we noticed that the "Boxes Sold" column was an `object` type because it had a couple of instances where it took the value "No Sales", therby preventing it from being read in as a numeric variable. We'd like to calculate some statistics based off the number of "Boxes Sold", so we'll need to convert "Boxes Sold" to a numeric variable type, we'll need to either replace these observations with a numeric value. 

Let's used what we learned about indexing to change the value of those "No Sales" observations to 0. Just as positional indexing can be used to change the value of specific objects within a list in Python, boolean indexing can be used to change tha value of specific observations in Pandas. For example, we can update the value of "Boxes Sold" by doing the following

In [96]:
sales_df.loc[sales_df["Boxes Sold"] == "No Sales", "Boxes Sold"] = 0

In the code above, we use boolean indexing to select just the rows we want to modify and which column we want to make changes to. We then set those observations to 0 by simply writing `= 0`. 

We can then print the dataset to see the updates

In [97]:
sales_df

Troop Number             Cookie Type Boxes Sold  \
0            179              Thin Mints         43   
1            179         Caramel deLites         16   
2            179   Peanut Butter Patties         22   
3            179      Girl Scout S'mores         11   
4            179               Lemonades         34   
5            179  Peanut Butter Sandwich          8   
6            179              Shortbread         38   
7            179            Thanks-A-Lot         26   
8            179  Caramel Chocolate Chip         17   
9            179           Toffee-tastic         29   
10           177              Thin Mints        117   
11           177         Caramel deLites         84   
12           177   Peanut Butter Patties         39   
13           177      Girl Scout S'mores         33   
14           177               Lemonades         56   
15           177  Peanut Butter Sandwich         12   
16           177              Shortbread         86   
17           177            Thanks-A-Lot         24   
18           177  Caramel Chocolate Chip         45   
19           177           Toffee-tastic         77   
20           177               Donations         20   
21           143              Thin Mints         94   
22           143         Caramel deLites         99   
23           143   Peanut Butter Patties          0   
24           143      Girl Scout S'mores         76   
25           143               Lemonades        125   
26           143  Peanut Butter Sandwich          0   
27           143              Shortbread         36   
28           143            Thanks-A-Lot         15   
29           143  Caramel Chocolate Chip         69   
30           143           Toffee-tastic        103   
31           212              Thin Mints         21   
32           212         Caramel deLites          7   
33           212   Peanut Butter Patties          3   
34           212      Girl Scout S'mores          5   
35           212               Lemonades         15   
36           212  Peanut Butter Sandwich          2   
37           212              Shortbread         17   
38           212            Thanks-A-Lot          9   
39           212  Caramel Chocolate Chip         12   
40           212           Toffee-tastic         13   
41           179              Thin Mints         43   
42           179         Caramel deLites         16   
43           179   Peanut Butter Patties         22   
44           179      Girl Scout S'mores         11   
45           179               Lemonades         34   
46           179  Peanut Butter Sandwich          8   
47           179              Shortbread         38   
48           179            Thanks-A-Lot         26   
49           179  Caramel Chocolate Chip         17   
50           179           Toffee-tastic         29   

                                    Notes  
0                                     NaN  
1                                     NaN  
2                                     NaN  
3                                     NaN  
4                                     NaN  
5                                     NaN  
6                                     NaN  
7                                     NaN  
8                                     NaN  
9                                     NaN  
10                                    NaN  
11                                    NaN  
12                                    NaN  
13                                    NaN  
14                                    NaN  
15                                    NaN  
16                                    NaN  
17                                    NaN  
18                                    NaN  
19                                    NaN  
20    *A kind soul just dropped off a $20  
21                                    NaN  
22                                    NaN  
23  *Severe Peanut Allergy Impacted Sales  
24                                    NaN  

### Now You Try!

A similar problem  as encountered above occured for the price data, `price_df`, that you created earlier. In this instance, one observation of the "Price (\\$)" variable was listed as "\\$5.00" rather than just 5, causing the column to be read-in as a string. So, in order to be able to covert this column to a numeric variable type, use boolean indexing to set the price of "Caramel Chocolate Chip" cookies to the numeric value `5`. Then print `price_df` to make sure your changes were implemented.

In [98]:
price_df.loc[price_df["Cookie"] == "Caramel Chocolate Chip", "Price ($)"] = 5

In [99]:
price_df

Cookie Price ($)
0               Thin Mints         4
1          Caramel deLites         4
2    Peanut Butter Patties         4
3       Girl Scout S'mores         4
4                Lemonades         4
5    Peanut Butter Sanwich         4
6               Shortbread         4
7             Thanks-A-Lot         4
8   Caramel Chocolate Chip         5
9                Lemonades         4
10           Toffee-tastic         4

### <ins>Change Column Data Types</ins>

Now that the "Boxes Sold" column in `sales_df` has only numeric values, we can change the column to a numeric data type, such as `int` or `float`. Since the data are all integers, we'll pick `int`.

In order to change the column type, we can use the [`.astype()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.astype.html) method. The `.astype()` method is a method associated with the Series object that converts a column of one type to another type based on a basic Python variable type.

In [100]:
sales_df["Boxes Sold"] = sales_df["Boxes Sold"].astype(int)

In [101]:
sales_df.dtypes

Troop Number     int64
Cookie Type     object
Boxes Sold       int32
Notes           object
dtype: object

In the code above, we use the `.astype()` method in combination with the `int` basic Python variable type to convert the "Boxes Sold" column to an integer.

We can also convert numeric columns into the `object` type. For example, we may want the "Troop Number" column to be a string since we don't plan on doing any arithmetic with the troop number. We can do that as below

In [102]:
sales_df["Troop Number"] = sales_df["Troop Number"].astype(str)
sales_df.dtypes

Troop Number    object
Cookie Type     object
Boxes Sold       int32
Notes           object
dtype: object

### Now You Try!

Convert the "Price (\\$)" column to an integer using the `.astype()` method. Then print the `.dtypes` of the `price_df` variable.

In [103]:
price_df["Price ($)"] = price_df["Price ($)"].astype(int)
price_df.dtypes

Cookie       object
Price ($)     int32
dtype: object

### <ins>Drop Duplicates</ins>

As we also saw above, `sales_df` has some duplicate observations. Luckily, working with duplicate observations is a simple task in Pandas. For example, we can quickly identify duplicate observations using the [`.duplicated()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) method, as below 

In [104]:
sales_df[sales_df.duplicated()]

Troop Number             Cookie Type  Boxes Sold Notes
41          179              Thin Mints          43   NaN
42          179         Caramel deLites          16   NaN
43          179   Peanut Butter Patties          22   NaN
44          179      Girl Scout S'mores          11   NaN
45          179               Lemonades          34   NaN
46          179  Peanut Butter Sandwich           8   NaN
47          179              Shortbread          38   NaN
48          179            Thanks-A-Lot          26   NaN
49          179  Caramel Chocolate Chip          17   NaN
50          179           Toffee-tastic          29   NaN

Note a couple things in the code above. First, the `duplicated()` method returns a boolean mask that identifies which observations are duplicates. In this case, `duplicated()` only identifes the second (or later) instance of an observation; it does not return all observations for which there are duplicates. Another way to think about this is that `duplicated()` only returns the excess or superfluous observations, i.e. the ones that would be dropped if you wished to drop duplicate observations. 

Also note that we introduced a new syntax here, in this case a shorthand for row indexing. In the above code, we supplied a boolean mask inside of square brackets adjoining our DataFrame and this returned just the rows associated we wanted to view. While this allows you to quickly view select rows, make sure to always use the `.loc` method when selecting specific rows *and* specific columns.

How do we drop these duplicates? We can easily do that with the [`drop_duplicates()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) method.

In [105]:
sales_df = sales_df.drop_duplicates()

In [106]:
sales_df

Troop Number             Cookie Type  Boxes Sold  \
0           179              Thin Mints          43   
1           179         Caramel deLites          16   
2           179   Peanut Butter Patties          22   
3           179      Girl Scout S'mores          11   
4           179               Lemonades          34   
5           179  Peanut Butter Sandwich           8   
6           179              Shortbread          38   
7           179            Thanks-A-Lot          26   
8           179  Caramel Chocolate Chip          17   
9           179           Toffee-tastic          29   
10          177              Thin Mints         117   
11          177         Caramel deLites          84   
12          177   Peanut Butter Patties          39   
13          177      Girl Scout S'mores          33   
14          177               Lemonades          56   
15          177  Peanut Butter Sandwich          12   
16          177              Shortbread          86   
17          177            Thanks-A-Lot          24   
18          177  Caramel Chocolate Chip          45   
19          177           Toffee-tastic          77   
20          177               Donations          20   
21          143              Thin Mints          94   
22          143         Caramel deLites          99   
23          143   Peanut Butter Patties           0   
24          143      Girl Scout S'mores          76   
25          143               Lemonades         125   
26          143  Peanut Butter Sandwich           0   
27          143              Shortbread          36   
28          143            Thanks-A-Lot          15   
29          143  Caramel Chocolate Chip          69   
30          143           Toffee-tastic         103   
31          212              Thin Mints          21   
32          212         Caramel deLites           7   
33          212   Peanut Butter Patties           3   
34          212      Girl Scout S'mores           5   
35          212               Lemonades          15   
36          212  Peanut Butter Sandwich           2   
37          212              Shortbread          17   
38          212            Thanks-A-Lot           9   
39          212  Caramel Chocolate Chip          12   
40          212           Toffee-tastic          13   

                                    Notes  
0                                     NaN  
1                                     NaN  
2                                     NaN  
3                                     NaN  
4                                     NaN  
5                                     NaN  
6                                     NaN  
7                                     NaN  
8                                     NaN  
9                                     NaN  
10                                    NaN  
11                                    NaN  
12                                    NaN  
13                                    NaN  
14                                    NaN  
15                                    NaN  
16                                    NaN  
17                                    NaN  
18                                    NaN  
19                                    NaN  
20    *A kind soul just dropped off a $20  
21                                    NaN  
22                                    NaN  
23  *Severe Peanut Allergy Impacted Sales  
24                                    NaN  
25                                    NaN  
26  *Severe Peanut Allergy Impacted Sales  
27                                    NaN  
28                                    NaN  
29                                    NaN  
30                                    NaN  
31                                    NaN  
32                                    NaN  
33                                    NaN  
34                                    NaN  
35                                    NaN  
36                                    NaN  
37                  

As we see above, the `.drop_duplicates()` method dropped the observations indicated by the `.duplicated()` method.

### Now You Try!

Use the `.duplicated()` and `.drop_duplicates()` method to first identify and then drop duplicates in `price_df`.

In [107]:
price_df[price_df.duplicated()]

Cookie  Price ($)
9  Lemonades          4

In [108]:
price_df = price_df.drop_duplicates()

In [109]:
price_df

Cookie  Price ($)
0               Thin Mints          4
1          Caramel deLites          4
2    Peanut Butter Patties          4
3       Girl Scout S'mores          4
4                Lemonades          4
5    Peanut Butter Sanwich          4
6               Shortbread          4
7             Thanks-A-Lot          4
8   Caramel Chocolate Chip          5
10           Toffee-tastic          4

### <ins>Dropping Observations and Columns</ins>

Lastly, we'll look at how to drop DataFrame observations and columns.

In `sales_df`, we can see that the Cookie Type of observation 20 is "Donations" and has an accompanying note that "A kind soul just dropped off a \\$20". Let's drop that observation so that the dataset is limited to only cookie sales. 

We can easily drop observations using boolean indexing as follows

In [110]:
sales_df = sales_df[sales_df["Cookie Type"] != "Donations"]

In the above code, we limit the data to just observations where "Cookie Type" doesn't equal "Donations". Then, we replace the existing `sales_df` DatFrame with that limited DataFrame to solidify the changes. 

For columns, we can either specify just the columns we intend to keep or the ones we intend to drop. To limit to just the columns we intend to keep, we can use the syntax we encountered earlier and pass a *list* of the columns we want to keep within the DataFrame indexer. For example, if we wanted to keep just "Cookie Type" and "Boxes Sold", we could say

```python
sales_df = sales_df[["Cookie Type", "Boxes Sold"]]
```

If we have many columns and only want to drop one or two of them, however, this method can be a bit arduous. In that instance, rather, we should use the [`.drop()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html) method to specify just the columns we'd want to drop. For example, in `sales_df`, we can go ahead and drop the "Notes" column by doing the following

In [111]:
sales_df = sales_df.drop(columns=["Notes"])

In the code above, we call the .`drop()` method and then pass a list of column names we want to drop to the named keyword argument `columns`. This drops the specified columns and the we assign the output to the variable `sales_df` to overwrite the DataFrame.

We can check that this worked by printing the list of the existing columns. We can do this in the Pandas using a new attribute we haven't encountered yet called the `.columns` attribute.

In [112]:
sales_df.columns

Index(['Troop Number', 'Cookie Type', 'Boxes Sold'], dtype='object')

As we seem the "Notes" columns has been dropped from the DataFrame. 

### Now You Try!

Use boolean indexing to limit `sales_df` to observations with positive non-zero values of "Boxes Sold"

In [113]:
sales_df = sales_df[sales_df["Boxes Sold"] > 0]

In [114]:
sales_df

Troop Number             Cookie Type  Boxes Sold
0           179              Thin Mints          43
1           179         Caramel deLites          16
2           179   Peanut Butter Patties          22
3           179      Girl Scout S'mores          11
4           179               Lemonades          34
5           179  Peanut Butter Sandwich           8
6           179              Shortbread          38
7           179            Thanks-A-Lot          26
8           179  Caramel Chocolate Chip          17
9           179           Toffee-tastic          29
10          177              Thin Mints         117
11          177         Caramel deLites          84
12          177   Peanut Butter Patties          39
13          177      Girl Scout S'mores          33
14          177               Lemonades          56
15          177  Peanut Butter Sandwich          12
16          177              Shortbread          86
17          177            Thanks-A-Lot          24
18          177  Caramel Chocolate Chip          45
19          177           Toffee-tastic          77
21          143              Thin Mints          94
22          143         Caramel deLites          99
24          143      Girl Scout S'mores          76
25          143               Lemonades         125
27          143              Shortbread          36
28          143            Thanks-A-Lot          15
29          143  Caramel Chocolate Chip          69
30          143           Toffee-tastic         103
31          212              Thin Mints          21
32          212         Caramel deLites           7
33          212   Peanut Butter Patties           3
34          212      Girl Scout S'mores           5
35          212               Lemonades          15
36          212  Peanut Butter Sandwich           2
37          212              Shortbread          17
38          212            Thanks-A-Lot           9
39          212  Caramel Chocolate Chip          12
40          212           Toffee-tastic          13

### <ins>Using Built-In Methods</ins>

There is one dataset we haven't read in yet called "scout_counts.csv" that gives us the number of scouts associated with each Troop. Let's read in and view that data below.

In [121]:
scout_url = "https://raw.githubusercontent.com/cra-international/Intro-to-Python/master/Intro%20to%20Python%20-%20II/scout_counts.csv"
scout_df = pd.read_csv(scout_url)

In [122]:
scout_df

Troop Name  Scouts
0  Troop 179       7
1  Troop 177      13
2  Troop 143      21
3  Troop 212       3

Here we see that we have the number of scouts in a column called "Scouts" and the troop in a column named "Troop Name". Ultimately, we'd like to merge this data with the sales data, but in order to do so we need to remove the `"Troop "` substring from the values in the "Troop Name" column. If we were trying to do this on one string, we might use string indexing or the string's built-in `.replace()` method to do this. Luckily, we are able to access vectorized versions of these operations using the [`.str`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.html?highlight=series%20str#pandas.Series.str) attribute as below.

In [125]:
scout_df["Indexed Col"] = scout_df["Troop Name"].str[-3:]
scout_df["Replaced Col"] = scout_df["Troop Name"].str.replace("Troop ", "")
scout_df

Troop Name  Scouts Indexed Col Replaced Col
0  Troop 179       7         179          179
1  Troop 177      13         177          177
2  Troop 143      21         143          143
3  Troop 212       3         212          212

We see above that the `.str` attribute we are able to apply [string indexing](https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html#indexing-with-str) and the string [`.replace()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.replace.html) method to every value in the "Troop Name" column.

Nearly every string method is accessible using the `.str` attribute. Although we won't touch on Datetime types, Datetime methods are accessible using the `.dt` attribute in Pandas.

In the code below, we'll drop one of the columns we just created and the other so we can use it for to merge the scout data with the sales data later on.

In [127]:
scout_df = scout_df.rename(columns={"Indexed Col" : "Troop Number"})
scout_df = scout_df.drop(columns=["Replaced Col"])

### Now You Try!

The "Cookie" column in `price_df` is all lower case. We'll also want to merge this data with `sales_df` so use the `.str` attribute to lowercase the values in the "Cookie Type" column in `sales_df`. 

In [ ]:
sales_df["Cookie Type"] = sales_df["Cookie Type"].str.lower()
sales_df.head()

## Part 6: Merging Data

Now that we have clean data, we are able to create more comprehensive datasets by merging datasets together. For example, we would like to merge the sales and price data together so that we can calculate the revenue associated with each product for each Troop. We can do that by using the [`.merge()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) method. The `.merge()` method has a number of keyword arguments that we will explore that allow you to determine exactly how the datasets are merged. We will show how to merge `sales_df` and `price_df`and then explain below.

In [115]:
merged_df = sales_df.merge(price_df, how="outer", left_on="Cookie Type", right_on="Cookie", indicator=True, validate="m:1")

In [116]:
merged_df

Troop Number             Cookie Type  Boxes Sold                  Cookie  \
0           179              Thin Mints        43.0              Thin Mints   
1           177              Thin Mints       117.0              Thin Mints   
2           143              Thin Mints        94.0              Thin Mints   
3           212              Thin Mints        21.0              Thin Mints   
4           179         Caramel deLites        16.0         Caramel deLites   
5           177         Caramel deLites        84.0         Caramel deLites   
6           143         Caramel deLites        99.0         Caramel deLites   
7           212         Caramel deLites         7.0         Caramel deLites   
8           179   Peanut Butter Patties        22.0   Peanut Butter Patties   
9           177   Peanut Butter Patties        39.0   Peanut Butter Patties   
10          212   Peanut Butter Patties         3.0   Peanut Butter Patties   
11          179      Girl Scout S'mores        11.0      Girl Scout S'mores   
12          177      Girl Scout S'mores        33.0      Girl Scout S'mores   
13          143      Girl Scout S'mores        76.0      Girl Scout S'mores   
14          212      Girl Scout S'mores         5.0      Girl Scout S'mores   
15          179               Lemonades        34.0               Lemonades   
16          177               Lemonades        56.0               Lemonades   
17          143               Lemonades       125.0               Lemonades   
18          212               Lemonades        15.0               Lemonades   
19          179  Peanut Butter Sandwich         8.0                     NaN   
20          177  Peanut Butter Sandwich        12.0                     NaN   
21          212  Peanut Butter Sandwich         2.0                     NaN   
22          179              Shortbread        38.0              Shortbread   
23          177              Shortbread        86.0              Shortbread   
24          143              Shortbread        36.0              Shortbread   
25          212              Shortbread        17.0              Shortbread   
26          179            Thanks-A-Lot        26.0            Thanks-A-Lot   
27          177            Thanks-A-Lot        24.0            Thanks-A-Lot   
28          143            Thanks-A-Lot        15.0            Thanks-A-Lot   
29          212            Thanks-A-Lot         9.0            Thanks-A-Lot   
30          179  Caramel Chocolate Chip        17.0  Caramel Chocolate Chip   
31          177  Caramel Chocolate Chip        45.0  Caramel Chocolate Chip   
32          143  Caramel Chocolate Chip        69.0  Caramel Chocolate Chip   
33          212  Caramel Chocolate Chip        12.0  Caramel Chocolate Chip   
34          179           Toffee-tastic        29.0           Toffee-tastic   
35          177           Toffee-tastic        77.0           Toffee-tastic   
36          143           Toffee-tastic       103.0           Toffee-tastic   
37          212           Toffee-tastic        13.0           Toffee-tastic   
38          NaN                     NaN         NaN   Peanut Butter Sanwich   

    Price ($)      _merge  
0         4.0        both  
1         4.0        both  
2         4.0        both  
3         4.0        both  
4         4.0        both  
5         4.0        both  
6         4.0        both  
7         4.0        both  
8         4.0        both  
9         4.0        both  
10        4.0        both  
11        4.0        both  
12        4.0        both  
13        4.0        both  
14        4.0        both  
15        4.0        both  
16        4.0        both  
17        4.0        both  
18        4.0        both  
19        NaN   left_only  
20        NaN   left_only  
21        NaN   left_only  
22        4.0        both  
23        4.0        both  
24        4.0        both  
25        4.0        both  
26        4.0        both  
27        4.0        both  
28        4.0        both  
29

Firstly, note that `.merge()` is called from `sales_df` and the first argument is the DataFrame to merge with, in this case `price_df`. We then specify how we'd like to merge the datasets using the `how` statement. There are four options: "left", "right", "outer", and "inner". Those familiar with SQL should be familiar with these. In this instance, the DataFrame from which the `merge()` method was called is the "left" dataset and the DataFrame specified inside the `merge()` statement is the "right" DataFrame. If you do a "left" merge by specifying `how="left"`, then all observations from the "left" DataFrame are kept and observations that don't merge from the "right" DataFrame are dropped. A "right" merge is the flipside of that. An "inner" merge, then, only keeps merged observations while an "outer" merge keeps all observations, regardless of where they are from or if they merged. 

Next, we specify the merge keys. Those familiar with Stata are used to the variables needing to be the same name prior to the merge. In Pandas, however, we can simply specify which column in the "left" DataFrame to use as the merge key and which column in the "right" DataFrame to use as its merge key. We can do that by specifying `left_on` and `right_on`, respectively. If the merging column is the same name in both DataFrames, then that column can be specified with the `on` argument and `left_on` and `right_on` no longer need to be specified.

Then, we specify `indicator=True` which creates a column called "\_merge" by default and *indicates* which dataset the observation originated in. If the observation is the result of a successful merge, it came from "both" datasets, otherwise it could come from the "left" DataFrame only or the "right" DataFrame. Stata users should be familiar with this syntax.

Finally, we specify validate="m:1" which checks that the merge key is "many" or non-unique in the "left" DataFrame and unique in the "right" DataFrame. If this is not the case, the function with throw an error. It is good practice to specify the `validate` argument as many-to-many merges are defined in Pandas which may lead to unexpected results. Again, users with some Stata experience should be familiar with this syntax.

It looks like some of the data from `price_df` didn't merge with `sales_df` because "Peanut Butter Sandwich" was mispelled in `price_df`. So, we can fix that and then rerun the merge and make sure all values merge as we expect.

In [117]:
price_df.loc[price_df["Cookie"] == "Peanut Butter Sanwich", "Cookie"] = "Peanut Butter Sandwich"
price_df

C:\Users\cwilson\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


Cookie  Price ($)
0               Thin Mints          4
1          Caramel deLites          4
2    Peanut Butter Patties          4
3       Girl Scout S'mores          4
4                Lemonades          4
5   Peanut Butter Sandwich          4
6               Shortbread          4
7             Thanks-A-Lot          4
8   Caramel Chocolate Chip          5
10           Toffee-tastic          4

In [118]:
merged_df = sales_df.merge(price_df, how="outer", left_on="Cookie Type", right_on="Cookie", indicator=True, validate="m:1")
merged_df

Troop Number             Cookie Type  Boxes Sold                  Cookie  \
0           179              Thin Mints          43              Thin Mints   
1           177              Thin Mints         117              Thin Mints   
2           143              Thin Mints          94              Thin Mints   
3           212              Thin Mints          21              Thin Mints   
4           179         Caramel deLites          16         Caramel deLites   
5           177         Caramel deLites          84         Caramel deLites   
6           143         Caramel deLites          99         Caramel deLites   
7           212         Caramel deLites           7         Caramel deLites   
8           179   Peanut Butter Patties          22   Peanut Butter Patties   
9           177   Peanut Butter Patties          39   Peanut Butter Patties   
10          212   Peanut Butter Patties           3   Peanut Butter Patties   
11          179      Girl Scout S'mores          11      Girl Scout S'mores   
12          177      Girl Scout S'mores          33      Girl Scout S'mores   
13          143      Girl Scout S'mores          76      Girl Scout S'mores   
14          212      Girl Scout S'mores           5      Girl Scout S'mores   
15          179               Lemonades          34               Lemonades   
16          177               Lemonades          56               Lemonades   
17          143               Lemonades         125               Lemonades   
18          212               Lemonades          15               Lemonades   
19          179  Peanut Butter Sandwich           8  Peanut Butter Sandwich   
20          177  Peanut Butter Sandwich          12  Peanut Butter Sandwich   
21          212  Peanut Butter Sandwich           2  Peanut Butter Sandwich   
22          179              Shortbread          38              Shortbread   
23          177              Shortbread          86              Shortbread   
24          143              Shortbread          36              Shortbread   
25          212              Shortbread          17              Shortbread   
26          179            Thanks-A-Lot          26            Thanks-A-Lot   
27          177            Thanks-A-Lot          24            Thanks-A-Lot   
28          143            Thanks-A-Lot          15            Thanks-A-Lot   
29          212            Thanks-A-Lot           9            Thanks-A-Lot   
30          179  Caramel Chocolate Chip          17  Caramel Chocolate Chip   
31          177  Caramel Chocolate Chip          45  Caramel Chocolate Chip   
32          143  Caramel Chocolate Chip          69  Caramel Chocolate Chip   
33          212  Caramel Chocolate Chip          12  Caramel Chocolate Chip   
34          179           Toffee-tastic          29           Toffee-tastic   
35          177           Toffee-tastic          77           Toffee-tastic   
36          143           Toffee-tastic         103           Toffee-tastic   
37          212           Toffee-tastic          13           Toffee-tastic   

    Price ($) _merge  
0           4   both  
1           4   both  
2           4   both  
3           4   both  
4           4   both  
5           4   both  
6           4   both  
7           4   both  
8           4   both  
9           4   both  
10          4   both  
11          4   both  
12          4   both  
13          4   both  
14          4   both  
15          4   both  
16          4   both  
17          4   both  
18          4   both  
19          4   both  
20          4   both  
21          4   both  
22          4   both  
23          4   both  
24          4   both  
25          4   both  
26          4   both  
27          4   both  
28          4   both  
29          4   both  
30          5   both  
31          5   both  
32          5   both  
33          5   both  
34          4   both  
35          4   both  
36          4   both  
37          4   both

Fantastic, it looks like everything merged successfully. Now, we can drop the extra key variable "Cookie" and the indicator "\_merge" so that we are left with a clean dataset.

In [119]:
merged_df = merged_df.drop(columns=["_merge", "Cookie"])
merged_df.columns

Index(['Troop Number', 'Cookie Type', 'Boxes Sold', 'Price ($)'], dtype='object')

### Now You Try!

Use "Troop Number" to merge `scout_df` with `merged_df` and create a new DataFrame, `analysis_df`. 